src/transport/raw/TCP.cpp

```diff
@@ -353,171 +439,243 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
-        endPoint->OnPeerClose          = OnPeerClosed;
+        // Update state for the active connection
+        activeConnection->Init(endPoint, addr);
+        tcp->mUsedEndPointCount++;
+        activeConnection->mConnectionState = TCPState::kConnected;
+
+        char addrStr[Transport::PeerAddress::kMaxToStringSize];
+        peerAddress.ToString(addrStr);
+        ChipLogProgress(Inet, "Incoming connection established with peer at %s.", addrStr);
+
+        // Call the upper layer handler for incoming connection received.
+        tcp->HandleConnectionReceived(activeConnection);
     }
     else
     {
-        ChipLogError(Inet, "Insufficient connection space to accept new connections");
+        ChipLogError(Inet, "Insufficient connection space to accept new connections.");
         endPoint->Free();
+        listenEndPoint->OnAcceptError(endPoint, CHIP_ERROR_TOO_MANY_CONNECTIONS);
     }
 }
 
-void TCPBase::OnAcceptError(Inet::TCPEndPoint * endPoint, CHIP_ERROR err)
+void TCPBase::HandleAcceptError(Inet::TCPEndPoint * endPoint, CHIP_ERROR err)
+{
+    endPoint->Free();
+    ChipLogError(Inet, "Accept error: %" CHIP_ERROR_FORMAT, err.Format());
+}
+
+CHIP_ERROR TCPBase::TCPConnect(const PeerAddress & address, Transport::AppTCPConnectionCallbackCtxt * appState,
+                               Transport::ActiveTCPConnectionState ** outPeerConnState)
 {
-    ChipLogError(Inet, "Accept error: %s", ErrorStr(err));
+    VerifyOrReturnError(mState == TCPState::kInitialized, CHIP_ERROR_INCORRECT_STATE);
+
+    // Verify that PeerAddress AddressType is TCP
+    VerifyOrReturnError(address.GetTransportType() == Transport::Type::kTcp, CHIP_ERROR_INVALID_ARGUMENT);
+
+    VerifyOrReturnError(mUsedEndPointCount < mActiveConnectionsSize, CHIP_ERROR_NO_MEMORY);
+
+    char addrStr[Transport::PeerAddress::kMaxToStringSize];
+    address.ToString(addrStr);
+    ChipLogProgress(Inet, "Connecting to peer %s.", addrStr);
+
+    ReturnErrorOnFailure(StartConnect(address, appState, outPeerConnState));
+
+    return CHIP_NO_ERROR;
 }
 
-void TCPBase::Disconnect(const PeerAddress & address)
+void TCPBase::TCPDisconnect(const PeerAddress & address)
 {
+    CHIP_ERROR err = CHIP_NO_ERROR;
     // Closes an existing connection
     for (size_t i = 0; i < mActiveConnectionsSize; i++)
     {
-        if (mActiveConnections[i].InUse())
+        if (mActiveConnections[i].IsConnected())
         {
             Inet::IPAddress ipAddress;
             uint16_t port;
             Inet::InterfaceId interfaceId;
 
-            mActiveConnections[i].mEndPoint->GetPeerInfo(&ipAddress, &port);
-            mActiveConnections[i].mEndPoint->GetInterfaceId(&interfaceId);
-            if (address == PeerAddress::TCP(ipAddress, port, interfaceId))
+            err = mActiveConnections[i].mEndPoint->GetPeerInfo(&ipAddress, &port);
+            if (err != CHIP_NO_ERROR)
+            {
+                ChipLogError(Inet, "GetPeerInfo error: %" CHIP_ERROR_FORMAT, err.Format());
```

Comment:
Followup: These error logs in this function should make it clearer that this is while trying to disconnect TCP.

Reply:
Prepended with TCPDisconnect: to make it clear.